### PR TITLE
Inject custom FastAPI dependencies into context object without overriding default context

### PR DESCRIPTION
**Pull request by tsmith023**
## Description
The `__get_context()` dependency injector has been renamed to `__get_context_getter()` and changed to incorporate the injection of a custom `context_getter()` dependency injector.

@Kludex wrote the solution to be in terms of `signature` so that the default and custom context getters can be merged and stored within `self.context_getter`. In this way, the method functions need not be altered by this change. If the `context_getter` object is `None` then a null-closure is passed to the injector instead.

## Types of Changes

- [ ] Core
- [x] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Documentation

## Issues Fixed or Closed by This PR

* #1493

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by botberry** [comment]
Hi, thanks for contributing to Strawberry 🍓!

We noticed that this PR is missing a `RELEASE.md` file. We use that to automatically do releases here on GitHub and, most importantly, to PyPI!

So as soon as this PR is merged, a release will be made 🚀.

Here's an example of `RELEASE.md`:

```markdown
Release type: patch

Description of the changes, ideally with some examples, if adding a new feature.
```

Release type can be one of patch, minor or major. We use [semver](https://semver.org/), so make sure to pick the appropriate type. If in doubt feel free to ask :)


Here's the tweet text: 
```
🆕 Release (next) is out! Thanks to Tommy Smith for the PR 👏

Get it here 👉 https://github.com/strawberry-graphql/strawberry/releases/tag/(next)
```

**Review by patrick91 — COMMENTED** [review]
Nice <3, can you also add a couple of tests? 😊

**Comment by tsmith023** [comment]
Apologies for sending so many commits and requests to run workflows. I didn't anticipate your testing harness being so comprehensive and opinionated, a lesson learnt for next time!

File: strawberry/fastapi/router.py
Comment (Kludex) [inline strawberry/fastapi/router.py:47]: It really doesn't matter if it's a positional parameter or _keyworded_ one, as you'll modify the signature below.

The [mutable default value](https://docs.quantifiedcode.com/python-anti-patterns/correctness/mutable_default_value_as_argument.html) is unnecessary on the dependency.

```suggestion
        async def dependency(
            custom_getter: Dict[str, Any],
            background_tasks: BackgroundTasks,
            request: Request = None,
            ws: WebSocket = None,
        ):
            return {
                "request": request or ws,
                "background_tasks": background_tasks,
                **custom_getter,
            }
```
The order also doesn't matter, jfyk.

File: strawberry/fastapi/router.py
Comment (Kludex) [inline strawberry/fastapi/router.py:55]: There's an alternative solution to basically build all the function. Something between those:

- https://github.com/Kludex/fastapi-authorization/blob/67765b28bd350bfef83288c4e4c36c55725f255a/fastapi_authorization/abac.py#L32-L45
- https://github.com/Kludex/fastapi-health/blob/f012f5a1e7a2e184fd7a51a17cf4c26215bb605e/fastapi_health/route.py#L103-L122 

But the current one is fine. :+1:

File: strawberry/fastapi/router.py
Comment (tsmith023) [reply strawberry/fastapi/router.py]: Wow, I never knew about that dictionary anti-pattern. Thanks for informing me!

Comment (Kludex) [reply strawberry/fastapi/router.py]: Be aware that your code would have never reached the antipattern behavior. I'm just mentioning because it is not really necessary, and it avoids future PRs about "fixing" what is actually not doing anything :kissing:

Comment (tsmith023) [reply strawberry/fastapi/router.py]: This change causes: `msg        = "duplicate parameter name: 'custom_getter'"` due to `parameters = [<Parameter "background_tasks: starlette.background.BackgroundTasks">, <Parameter "custom_getter: Dict[str, Any]">, <P...eter "request: starlette.requests.Request = None">, <Parameter "custom_getter: Dict[str, Any] = Depends(get_context)">]`, so a default value is required. I'll set it to `None` though to avoid that anti-pattern.

Comment (tsmith023) [reply strawberry/fastapi/router.py]: There were errors caused by this: `parameters = [<Parameter "background_tasks: starlette.background.BackgroundTasks">, <Parameter "custom_getter: Dict[str, Any]">, <P...eter "request: starlette.requests.Request = None">, <Parameter "custom_getter: Dict[str, Any] = Depends(get_context)">]`, so line 57 of `router.py` should read
```python
sig.parameters["custom_getter: Dict[str, Any]"].replace(default=Depends(custom_getter))`.
```

Is this unusual that the type signatures are being supplied to `signature()` by `dependency()`?

Comment (Kludex) [reply strawberry/fastapi/router.py]: ups! I'm sorry, the order matters because we're pasting the n - 1 parameters back to the signature :sweat_smile: 

image

Comment (tsmith023) [reply strawberry/fastapi/router.py]: Aha, that's what that cryptic line is doing! Your method makes sense to me now, thanks for the help :)

File: strawberry/fastapi/router.py
Comment (BryceBeagle) [inline strawberry/fastapi/router.py:90]:
```suggestion
        self.context_getter = self.__get_context_getter(context_getter or lambda: {})
```

File: strawberry/fastapi/router.py
Comment (BryceBeagle) [inline strawberry/fastapi/router.py:42]:
```suggestion
    def __get_context_getter(
        custom_getter: Callable[..., Optional[Dict[str, Any]]]
    ) -> Callable[..., Dict[str, Any]]:
        async def dependency(
            custom_getter: Dict[str, Any],
            background_tasks: BackgroundTasks,
            request: Optional[Request] = None,
            ws: Optional[WebSocket] = None,
        ) -> Callable[..., Dict[str, Any]]:
```

Comment (BryceBeagle) [reply strawberry/fastapi/router.py:42]: Can the `Any`s here be narrowed to `object`s?

File: strawberry/fastapi/router.py
Comment (BryceBeagle) [reply strawberry/fastapi/router.py:52]: If you could add a little comment explaining why the first item in the dict is dropped, that'd be good too. It's not immediately obvious

**Review by BryceBeagle — CHANGES_REQUESTED** [review]
Some minor suggestions/requests :+1:

File: strawberry/fastapi/router.py
Comment (tsmith023) [reply strawberry/fastapi/router.py:42]: Do you mean the `Any`s provided to `custom_getter` or just those within the `Callable[]`s?

File: strawberry/fastapi/router.py
Comment (tsmith023) [reply strawberry/fastapi/router.py:90]: I tried this before and was getting a syntax error. However, looking at it again with fresh eyes, it works if the `lambda` is enclosed in parentheses, so:
```python
self.context_getter = self.__get_context_getter(context_getter or (lambda: {}))
```

File: strawberry/fastapi/router.py
Comment (tsmith023) [reply strawberry/fastapi/router.py]: The `Optional`s here trigger typing errors within the fastAPI testing suite so can't be included. Is it an issue if these types are omitted since they have default values of `None`?

Comment (Kludex) [reply strawberry/fastapi/router.py]: I don't think you have an option.

File: strawberry/fastapi/router.py
Comment (BryceBeagle) [reply strawberry/fastapi/router.py]: Now that I think about it, would
```python
self.context_getter = self.__get_context_getter(context_getter or dict)
```
work?

Comment (tsmith023) [reply strawberry/fastapi/router.py]: It needs to be a lambda closure as `context_getter` is a `Callable` that injects dependencies. However, I've also decided it's easier to have `context_getter or (lambda: None)` here and then `**(custom_getter or {})` on line 46. These are all requirements for the MyPy checks to pass.

**Comment by tsmith023** [comment]
I think that this is in a finalised state now that can be reviewed. I have it passing the test suite on my local machine and working within my projects so it just needs the sign-off of any or all of @patrick91, @Kludex, @BryceBeagle :)

File: strawberry/fastapi/router.py
Comment (tsmith023) [reply strawberry/fastapi/router.py:55]: > Suggestion, but you can ignore if you don't like it. Benefit of `itertools.islice` here is pretty minimal:

I feel like leaving this as is since, as you say, the benefit is minimal and appears less 'pythonic' to me

**Review by BryceBeagle — APPROVED** [review]

File: RELEASE.md
Comment (patrick91) [inline RELEASE.md:3]:
```suggestion
Change `context_getter` in `strawberry.fastapi.GraphQLRouter` to merge, rather than overwrite, default and custom getters.

This mean now you can always access the `request` instance from `info.context`, even when using
a custom context getter.
```

**Review by patrick91 — APPROVED** [review]
Awesome! Thanks you 😊

**Comment by botberry** [comment]
Thanks for contributing to Strawberry! 🎉 You've been invited to join
the Strawberry GraphQL organisation 😊

You can also request a free sticker by filling this form: https://forms.gle/dmnfQUPoY5gZbVT67

And don't forget to join our discord server: https://strawberry.rocks/discord 🔥

**Pull request merged**
